internal/archiver/archiver.go

```diff
@@ -615,6 +638,72 @@ func (arch *Archiver) save(ctx context.Context, snPath, target string, previous
 	return fn, false, nil
 }
 
+// Given a symlink, return the absolute path of its target target.
+// Note: unlike `filepath.EvalSymlinks`, this does not recurse! If the target
+// is itself a symlink, we just return the target.
+func (arch *Archiver) resolveSymlink(symlink string) (string, error) {
+	target, err := os.Readlink(symlink)
+	if err != nil {
+		return "", err
+	}
+
+	if filepath.IsAbs(target) {
+		return target, nil
+	}
+
+	// If the filepath is relative, then resolve it relative to the directory
+	// of the symlink.
+	symlinkDir := filepath.Dir(symlink)
+	absTarget := filepath.Join(symlinkDir, target)
+	absTarget = filepath.Clean(absTarget)
+
+	return absTarget, nil
+}
+
+func (arch *Archiver) isSymlinkInSnapshot(symlink string) (bool, error) {
+	target, err := arch.resolveSymlink(symlink)
+	if err != nil {
+		return false, err
+	}
+
+	inSnapshot, err := arch.isFileInSnapshot(target)
+	if err != nil {
+		return false, err
+	}
+	if !inSnapshot {
+		return false, errors.Errorf("encountered a symlink pointing to a file that is not included in the backup: %s", target)
+	}
+
+	// If `target` is itself a symlink, verify that it's also in the snapshot.
+	fi, err := fs.Lstat(target)
+	if err != nil {
+		return false, err
+	}
+	if fi.Mode()&os.ModeSymlink > 0 {
+		return arch.isSymlinkInSnapshot(target)
+	}
+
+	return true, nil
+}
+
+func (arch *Archiver) isFileInSnapshot(file string) (bool, error) {
+	for _, snapshotTarget := range *arch.snapshotTargets {
+		relativePath, err := filepath.Rel(snapshotTarget, file)
+		if err != nil {
+			return false, fmt.Errorf("error computing relative path from %s to %s: %s", snapshotTarget, file, err)
+		}
+
+		if strings.HasPrefix(relativePath, "..") {
+			continue
+		}
+
+		// <<< TODO: look at `arch.SelectByName` and `arch.Select` >>>
```

Comment:
The basic idea of this PR is that every time we hit a symlink, check if it is (or will be) included in the snapshot. I'm not in love with this approach because it means that I'm effectively re-implementing the filtering logic in `arch.save`. If we add new types of filters in the future, it's possible the implementations could drift.

An alternative implementation that does not have this brittleness would be to instead record a "seen list" of all symlinks we encountered during the backup. Then, at the very end, confirm that each of those symlinks is actually in the snapshot. One downside of this is that there's no guaranteed bound on how large the "seen list" gets. I'm not familiar enough with restic's codebase to know if it's straightforward to walk through the snapshot, or if that's a no-no because it may be on some remote server and is expensive (in time or $) to walk through it.

A third implementation would be to wait until the very end to do the check. Walk through the whole snapshot, and for each symlink, verify that it's in the snapshot. This would mean a second pass through the whole snapshot. I have the same question here about if it's OK to walk through the snapshot on the end, or if it's not OK to do because it may be remote.

After typing this up, I think I prefer the "third implementation" the best. Would really appreciate advice from a restic maintainer, though!